presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java

```diff
@@ -0,0 +1,253 @@
+                    joinNode.getType(),
+                    new ProjectNode(planNodeIdAllocator.getNextId(), rewrittenLeft, leftAssignments.build()),
+                    new ProjectNode(planNodeIdAllocator.getNextId(), rewrittenRight, rightAssignments.build()),
+                    joinClauseBuilder.build(),
+                    joinOutputBuilder.build(),
+                    joinNode.getFilter(),
+                    joinNode.getLeftHashVariable(),
+                    joinNode.getRightHashVariable(),
+                    joinNode.getDistributionType(),
+                    joinNode.getDynamicFilters());
+        }
+
+        private RowExpression randomizeJoinKey(RowExpression keyExpression, String prefix)
+        {
+            int partitionCount = getHashPartitionCount(session);
+            RowExpression randomNumber = call(
+                    functionAndTypeManager,
+                    "random",
+                    BIGINT,
+                    constant((long) partitionCount, BIGINT));
+            RowExpression randomNumberVarchar = call("CAST", functionAndTypeManager.lookupCast(CAST, randomNumber.getType(), VARCHAR), VARCHAR, randomNumber);
+            RowExpression concatExpression = call(functionAndTypeManager,
+                    "concat",
+                    VARCHAR,
+                    ImmutableList.of(constant(Slices.utf8Slice(prefix), VARCHAR), randomNumberVarchar));
+
+            RowExpression castToVarchar = keyExpression;
+            // Only do cast if keyExpression is not VARCHAR type.
+            if (!(keyExpression.getType() instanceof VarcharType)) {
+                castToVarchar = call("CAST", functionAndTypeManager.lookupCast(CAST, keyExpression.getType(), VARCHAR), VARCHAR, keyExpression);
+            }
+            return new SpecialFormExpression(SpecialFormExpression.Form.COALESCE, VARCHAR, ImmutableList.of(castToVarchar, concatExpression));
+        }
+
+        // Do not need to generate randomized variable if the joinKey 1) has already been randomized with the same prefix 2) included in the output of the source node
+        private boolean isAlreadyRandomized(PlanNode source, VariableReferenceExpression joinKey, String prefix)
+        {
+            return keyToRandomKeyMap.containsKey(prefix) && keyToRandomKeyMap.get(prefix).containsKey(joinKey) && source.getOutputVariables().contains(keyToRandomKeyMap.get(prefix).get(joinKey));
+        }
+
+        private Map<VariableReferenceExpression, RowExpression> generateRandomKeyMap(List<VariableReferenceExpression> joinKeys, String prefix)
+        {
+            List<RowExpression> randomExpressions = joinKeys.stream().map(x -> randomizeJoinKey(x, prefix)).collect(ImmutableList.toImmutableList());
+            List<VariableReferenceExpression> randomVariable = randomExpressions.stream()
+                    .map(x -> planVariableAllocator.newVariable(x, RandomizeNullKeyInOuterJoin.class.getSimpleName()))
```

Comment:
Since this is generated by optimization rules, I intentionally name it with this hint so that it can stand out and easy to identify when debugging query plans.